### PR TITLE
adding mixed precision training, which improves throughput by about …

### PR DESCRIPTION
**Pull request by vinhngx**
…20-40% on GPUs with tensor cores

### CODE REVIEW

**Comment by vinhngx** [comment]
Please, could any maintainer of this repo helps review this?

**Comment by vinhngx** [comment]
@hangzhaomit?

**Comment by seedlite** [comment]
@vinhngx This seems interesting. How much gain in training time were you able to achieve by using mixed precision training? 
Also, do we have to make any changes to the test.py also while running inferences?

**Comment by seedlit** [comment]
@vinhngx Have you also prepared a script with mixed precision support for running inferences?

**Comment by vinhngx** [comment]
> 
> 
> @vinhngx This seems interesting. How much gain in training time were you able to achieve by using mixed precision training?
> Also, do we have to make any changes to the test.py also while running inferences?

@ghost though the training is done with FP16, the weight is stored as FP32 and there is no changed required when doing inference (in FP32)

**Comment by vinhngx** [comment]
> 
> 
> @vinhngx Have you also prepared a script with mixed precision support for running inferences?

I have not produced a script for mixed precision inference here, though for serious productization, converting the model to ONNX then accelerated with [TensorRT](https://developer.nvidia.com/tensorrt) is recommended.

**Comment by seedlit** [comment]
@vinhngx I am using this script for mixed precision training (on multiple gpus). However, it dosen't work well with O2, and O3. See [#227](https://github.com/NVIDIA/apex/issues/227). For O1, my gpus memory usage increases compared to FP32 training, and training time also increases. It seems we have to [wait a little](https://github.com/NVIDIA/apex/issues/269#issuecomment-566841593) till mixed precision training support is introduced in Pytorch 1.5

I am also getting 'overflowing gradients' when training on O1 (when setting batch_size_per_gpu = 1, for batch_size_per_gpu >1, it works fine), even after 
```
with amp.scale_loss(loss1, optimizer1) as scaled_loss:
    scaled_loss.backward()
```
Do you know any workaround this?

**Comment by JulianJuaner** [comment]
I am working on acceleration using mixed-precision training. Actually, it works fine when using a single GPU (about 8x faster than using 4 cards with the same batch size, half memory used). But this does not work for multiple GPUs. I think it is majorly caused by the new DataParallel and BatchNormSync implementation.
Or is it possible to modify the code to have a better performance on multiple GPUs?

**Comment by JulianJuaner** [comment]
> I am working on acceleration using mixed-precision training. Actually, it works fine when using a single GPU with several lines modification (about 8x faster than using 4 cards with the same batch size, half memory used). But this does not work for multiple GPUs. I think it is majorly caused by the new DataParallel and BatchNormSync implementation.
> Or is it possible to modify the code to have a better performance on multiple GPUs?
After referencing hszhao's DDP implementation on [this link](https://github.com/hszhao/semseg/blob/master/tool/train.py). Now my program has half GPU memory usage with 3x faster speed during the training.
